src/noshow/api/app_helpers.py

```diff
@@ -119,35 +134,31 @@ def create_treatment_groups(
             how="left",
         )
     else:
-        predictions["treatment_group"] = None
-
-    # Create prediction score bins using quantile bins
-    predictions["score_bin"] = pd.qcut(predictions["prediction"], q=n_bins)
-    predictions = predictions.sort_values(["prediction"])
-
-    # only keep top prediction per patient for treatment/control split
-    deduplicated = predictions.drop_duplicates(subset="pseudo_id", keep="first")
-    # only keep patients without treatment group
-    deduplicated = deduplicated[deduplicated["treatment_group"].isnull()]
-
-    # Create stratified randomization in control and treatment groups
-    deduplicated = deduplicated.sort_values(["hoofdagenda", "prediction"])
-    deduplicated["treatment_group"] = deduplicated.groupby(
-        ["hoofdagenda", "score_bin"], observed=True
-    )["prediction"].transform(lambda x: np.arange(len(x)) % 2)
-
-    # update predictions with treatment groups from deduplicated based on pseudo_id
-    merged = predictions.merge(
-        deduplicated[["pseudo_id", "treatment_group"]],
-        on="pseudo_id",
-        how="left",
-        suffixes=("", "_new"),
+        predictions.loc[:, "treatment_group"] = None
+    predictions = predictions.sort_values("prediction", ascending=False)
+    # apply bins based on supplied fixed score_bins
+    predictions = (
+        predictions.groupby("hoofdagenda")
+        .apply(apply_bins, bin_dict=bin_edges, include_groups=False)
+        .reset_index()
+    )
+    predictions = predictions.drop(columns="level_1")
+
+    # Fill NaN values in 'treatment_group' with calculated values
+    mask = predictions["treatment_group"].isnull()
+    predictions.loc[mask, "treatment_group"] = (
+        predictions[mask]
+        .groupby(["hoofdagenda", "score_bin"])["prediction"]
+        .transform(lambda x: (np.arange(len(x)) + random.randint(0, 1)) % 2)
     )
 
-    # Update the treatment_group values where there's a new value
-    predictions["treatment_group"] = merged["treatment_group_new"].combine_first(
-        predictions["treatment_group"]
+    # For every new patient assign the treatment group based on the mode
+    predictions.loc[mask, "treatment_group"] = (
```

Comment:
Als ik het goed begrijp gebeurt dit zodat als een patiënt voor het eerst een voorspelling krijgt en er meerdere afspraken op die dag zijn, de patiënt toch maar 1 groep geassigned krijgt? Misschien is het netter om dit voor de vorige stap af te handelen door alleen de eerste afspraken te assignen oid?

Reply:
besproken